batcher/aligned-batcher/src/lib.rs

```diff
@@ -1272,55 +1296,102 @@ impl Batcher {
+                user_balance_is_unlocked_retryable(
+                    &self.payment_service,
+                    &self.payment_service_fallback,
+                    addr,
+                )
+            },
+            DEFAULT_MIN_DELAY,
+            DEFAULT_FACTOR,
+            DEFAULT_MAX_TIMES,
+        )
+        .await
         {
-            return unlock_block != U256::zero();
+            Ok(result) => result,
+            Err(_) => {
+                warn!("Could not get user locking state.");
+                false
+            }
         }
-        warn!("Could not get user locking state");
-        false
     }
 
-    /// Gets the current gas price from Ethereum.
+    /// Gets the current gas price from Ethereum using exponential backoff.
     /// Returns `None` if the gas price couldn't be returned
-    /// FIXME: This should return a `Result` instead.
-    async fn get_gas_price(&self) -> Option<U256> {
-        if let Ok(gas_price) = self
-            .eth_ws_provider
-            .get_gas_price()
-            .await
-            .inspect_err(|e| warn!("Failed to get gas price. Trying with fallback: {e:?}"))
-        {
-            return Some(gas_price);
-        }
+    async fn get_gas_price(&self) -> Result<U256, BatcherError> {
+        retry_function(
+            || get_gas_price_retryable(&self.eth_ws_provider, &self.eth_ws_provider_fallback),
+            DEFAULT_MIN_DELAY,
+            DEFAULT_FACTOR,
+            DEFAULT_MAX_TIMES,
+        )
+        .await
+        .map_err(|e| {
+            error!("Could't get gas price: {e}");
+            BatcherError::GasPriceError
+        })
+    }
 
-        self.eth_ws_provider_fallback
-            .get_gas_price()
+    /// Uploads the batch to s3 using exponential backoff.
+    async fn upload_batch_to_s3(
+        &self,
+        batch_bytes: &[u8],
+        file_name: &str,
+    ) -> Result<(), BatcherError> {
+        retry_function(
+            || {
+                Self::upload_batch_to_s3_retryable(
+                    batch_bytes,
+                    file_name,
+                    self.s3_client.clone(),
+                    &self.s3_bucket_name,
+                )
+            },
+            DEFAULT_MIN_DELAY,
+            DEFAULT_FACTOR,
+            DEFAULT_MAX_TIMES,
+        )
+        .await
+        .map_err(|_| BatcherError::BatchUploadError("Error uploading batch to s3".to_string()))
+    }
+
+    async fn upload_batch_to_s3_retryable(
+        batch_bytes: &[u8],
+        file_name: &str,
+        s3_client: S3Client,
+        s3_bucket_name: &str,
+    ) -> Result<(), RetryError<String>> {
+        s3::upload_object(&s3_client, s3_bucket_name, batch_bytes.to_vec(), file_name)
             .await
-            .inspect_err(|e| warn!("Failed to get gas price: {e:?}"))
-            .ok()
+            .map_err(|e| {
+                warn!("Error uploading batch to s3 {e}");
+                RetryError::Transient(e.to_string())
+            })?;
```

Comment:
I wonder whether all errors are actually transient for S3.

Reply:
I guess it doesn't matter if we have a `MaxRetry` right?

Reply:
It depends on whether the delay matters. Failing fast might be a good thing sometimes.

Reply:
If having `MaxRetry` invalidates Permanent errors, there is no use case for Permanent errors.
We are missing the differentiation of Transient and Permanent errors in this PR, but it could be added in a later issue

Reply:
I'm not opposed at all to adding it later. I was mostly wondering if all errors were recoverable (it can be the case for some protocols I guess) and this was intentional or an oversight.